managers/bosh-manager/DirectorService.js

```diff
@@ -778,7 +768,13 @@ class DirectorService extends BaseDirectorService {
     return Promise.join(
         this.executeActions(CONST.SERVICE_LIFE_CYCLE.PRE_BIND, actionContext),
         this.getDeploymentIps(deploymentName),
-        (preBindResponse, ips) => this.agent.createCredentials(ips, binding.parameters, preBindResponse)
+        (preBindResponse, ips) => utils.retry(() => this.agent.createCredentials(ips, binding.parameters, preBindResponse), {
+          maxAttempts: 3,
+          timeout: CONST.AGENT.OPERATION_TIMEOUT_IN_MILLIS
```

Comment:
Do you want to add a predicate here to retry to only retry in case of TimeOut? You seem to be retrying even for other generic errors which Agent might be throwing.

Reply:
Can be better to retry as scenarios might be getinfo successful but createcreds not. Or there might be other unknown scenarios. Also, the timeout happens in 35 sec much lower than cf timeout. Is that fine? In all other retries in code base we don't check only for timeout. So, shall we keep the same for this as well ?

Reply:
You are retrying on the method 'createCredentials' - this method internally does two things - get ip (via /info) & create creds. Now if either of these fails, the method 'createCredentials' fails. Also when you are retrying in scenarios like above (especially creating sth), its better to retry on known scenarios. If unknown scenarios occur, which also qualifies for retry then they must be added in subsequently as known scenarios. Should not retry for anything and everything.

Reply:
With regards to this timeout value, better get this clarified once from service team if this SLA is acceptable to services teams. Also i thought you were a big proponent of configs over constant :-), if this is sth thats configurable wouldnt you want it in your configs ?

Reply:
+1 for keeping it a config with a default in the spec.

Reply:
sure, will put in config :)

Reply:
@jagadish-kb @sauravmndl  I feel retrying for all error scenarios is not going to create any issue, if not make it better. Finding the set of errors for which we need to retry is tricky, as errors like "credentials cannot be created as cluster topology changed in between" may also be types of error for which retry is necessary. I do not see any error for which retry should not be done, do you see any? As it is only be retried for finite amount of time, I guess it is better to keep the retry for all errors. Let me know what you think.

Reply:
@jagadish-kb As discussed keep as it is. Better to find out all possible errors and do retry only for those errors. So, not retrying only for timeout error. Anyway it will timeout in 35 secs (configurable now). In future if required we will find out list of all possible errors and do retry for only those for all agent operations.